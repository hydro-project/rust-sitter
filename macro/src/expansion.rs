use std::collections::HashSet;

use proc_macro2::Span;
use quote::ToTokens;
use rust_sitter_common::*;
use syn::{parse::Parse, punctuated::Punctuated, *};

fn is_sitter_attr(attr: &Attribute) -> bool {
    let ident = &attr.path.segments.iter().next().unwrap().ident;
    ident == "rust_sitter"
}

enum ParamOrField {
    Param(Expr),
    Field(FieldValue),
}

impl ToTokens for ParamOrField {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        match self {
            ParamOrField::Param(expr) => expr.to_tokens(tokens),
            ParamOrField::Field(field) => field.to_tokens(tokens),
        }
    }
}

fn gen_field(path: String, leaf: Field, out: &mut Vec<Item>) {
    let extract_ident = Ident::new(&format!("extract_{}", path), Span::call_site());
    let leaf_type = leaf.ty;

    let leaf_text_expr: Expr = syn::parse_quote! {
        node.and_then(|n| n.utf8_text(source).ok())
    };

    let leaf_attr = leaf
        .attrs
        .iter()
        .find(|attr| attr.path == syn::parse_quote!(rust_sitter::leaf));

    let leaf_params = leaf_attr.and_then(|a| {
        a.parse_args_with(Punctuated::<NameValueExpr, Token![,]>::parse_terminated)
            .ok()
    });

    let transform_param = leaf_params.as_ref().and_then(|p| {
        p.iter()
            .find(|param| param.path == "transform")
            .map(|p| p.expr.clone())
    });

    let mut skip_over = HashSet::new();
    skip_over.insert("Spanned");
    skip_over.insert("Box");

    let (inner_type, is_vec) = try_extract_inner_type(&leaf_type, "Vec", &skip_over);
    let (inner_type, is_option) = try_extract_inner_type(&inner_type, "Option", &skip_over);

    let (leaf_stmts, leaf_expr): (Vec<Stmt>, Expr) = match transform_param {
        Some(closure) => {
            if is_vec {
                panic!("Vec or Spanned of leaves is not supported");
            }

            (
                vec![syn::parse_quote! {
                    fn make_transform() -> impl Fn(&str) -> #inner_type {
                        #closure
                    }
                }],
                if is_option {
                    syn::parse_quote! {
                        #leaf_text_expr.map(|t| make_transform()(t))
                    }
                } else {
                    syn::parse_quote! {
                        make_transform()(#leaf_text_expr.unwrap())
                    }
                },
            )
        }
        None => {
            let element_field = format!("{}_vec_element", path);

            (
                vec![],
                syn::parse_quote!(rust_sitter::Extract::extract(node, source, Some(#element_field))),
            )
        }
    };

    out.push(syn::parse_quote! {
        #[allow(non_snake_case)]
        #[allow(clippy::unused_unit)]
        fn #extract_ident(node: Option<rust_sitter::Node>, source: &[u8]) -> #leaf_type {
            #(#leaf_stmts)*
            #leaf_expr
        }
    });
}

fn gen_struct_or_variant(
    path: String,
    fields: Fields,
    variant_ident: Option<Ident>,
    containing_type: Ident,
    out: &mut Vec<Item>,
) {
    fields.iter().enumerate().for_each(|(i, field)| {
        let ident_str = field
            .ident
            .as_ref()
            .map(|v| v.to_string())
            .unwrap_or(format!("{}", i));
        gen_field(
            format!("{}_{}", path.clone(), ident_str),
            field.clone(),
            out,
        );
    });

    let extract_ident = Ident::new(&format!("extract_{}", path), Span::call_site());

    let mut have_named_field = false;

    let children_parsed = fields
        .iter()
        .enumerate()
        .map(|(i, field)| {
            let ident_str = field
                .ident
                .as_ref()
                .map(|v| v.to_string())
                .unwrap_or(format!("{}", i));
            let ident = Ident::new(
                &format!("extract_{}_{}", path.clone(), ident_str),
                Span::call_site(),
            );

            let mut skip_over = HashSet::new();
            skip_over.insert("Spanned");
            skip_over.insert("Box");

            let expr = syn::parse_quote! {
                #ident(node.child_by_field_name(#ident_str), source)
            };

            if field.ident.is_none() {
                ParamOrField::Param(expr)
            } else {
                let field_name = field.ident.as_ref().unwrap();
                have_named_field = true;
                ParamOrField::Field(FieldValue {
                    attrs: vec![],
                    member: Member::Named(field_name.clone()),
                    colon_token: Some(Token![:](Span::call_site())),
                    expr,
                })
            }
        })
        .collect::<Vec<ParamOrField>>();

    if let Some(variant_ident) = variant_ident {
        if have_named_field {
            out.push(syn::parse_quote! {
                #[allow(non_snake_case)]
                fn #extract_ident(node: rust_sitter::Node, source: &[u8]) -> #containing_type {
                    #containing_type::#variant_ident {
                        #(#children_parsed),*
                    }
                }
            });
        } else {
            out.push(syn::parse_quote! {
                #[allow(non_snake_case)]
                fn #extract_ident(node: rust_sitter::Node, source: &[u8]) -> #containing_type {
                    #containing_type::#variant_ident(
                        #(#children_parsed),*
                    )
                }
            });
        }
    } else {
        out.push(syn::parse_quote! {
            #[allow(non_snake_case)]
            fn #extract_ident(node: rust_sitter::Node, source: &[u8]) -> #containing_type {
                #containing_type {
                    #(#children_parsed),*
                }
            }
        });
    }
}

pub fn expand_grammar(input: ItemMod) -> ItemMod {
    let grammar_name = input
        .attrs
        .iter()
        .find_map(|a| {
            if a.path == syn::parse_quote!(rust_sitter::grammar) {
                let grammar_name_expr = a.parse_args_with(Expr::parse).ok();
                if let Some(Expr::Lit(ExprLit {
                    attrs: _,
                    lit: Lit::Str(s),
                })) = grammar_name_expr
                {
                    Some(s.value())
                } else {
                    panic!("Expected string literal for grammar name");
                }
            } else {
                None
            }
        })
        .expect("Each grammar must have a name");

    let (brace, new_contents) = input.content.unwrap();

    let root_type = new_contents
        .iter()
        .find_map(|item| match item {
            Item::Enum(ItemEnum { ident, attrs, .. })
            | Item::Struct(ItemStruct { ident, attrs, .. }) => {
                if attrs
                    .iter()
                    .any(|attr| attr.path == syn::parse_quote!(rust_sitter::language))
                {
                    Some(ident.clone())
                } else {
                    None
                }
            }
            _ => None,
        })
        .expect("Each parser must have the root type annotated with `#[rust_sitter::language]`");

    let mut transformed: Vec<Item> = new_contents
        .iter()
        .cloned()
        .flat_map(|c| match c {
            Item::Enum(mut e) => {
                let mut impl_body = vec![];
                e.variants.iter().for_each(|v| {
                    gen_struct_or_variant(
                        format!("{}_{}", e.ident, v.ident),
                        v.fields.clone(),
                        Some(v.ident.clone()),
                        e.ident.clone(),
                        &mut impl_body,
                    )
                });

                let match_cases: Vec<Arm> = e
                    .variants
                    .iter()
                    .map(|v| {
                        let variant_path = format!("{}_{}", e.ident, v.ident);
                        let extract_ident =
                            Ident::new(&format!("extract_{}", variant_path), Span::call_site());
                        syn::parse_quote! {
                            #variant_path => #extract_ident(node.child(0).unwrap(), source)
                        }
                    })
                    .collect();

                e.attrs.retain(|a| !is_sitter_attr(a));
                e.variants.iter_mut().for_each(|v| {
                    v.attrs.retain(|a| !is_sitter_attr(a));
                    v.fields.iter_mut().for_each(|f| {
                        f.attrs.retain(|a| !is_sitter_attr(a));
                    });
                });

                let enum_name = &e.ident;
                let extract_impl: Item = syn::parse_quote! {
                    impl rust_sitter::Extract for #enum_name {
                        #[allow(non_snake_case)]
                        fn extract(node: Option<rust_sitter::Node>, source: &[u8], _vec_field_name: Option<&str>) -> Self {
                            let node = node.unwrap();
                            #(#impl_body)*
                            match node.child(0).unwrap().kind() {
                                #(#match_cases),*,
                                _ => panic!()
                            }
                        }
                    }
                };

                vec![Item::Enum(e), extract_impl]
            }

            Item::Struct(mut s) => {
                let mut impl_body = vec![];

                gen_struct_or_variant(
                    s.ident.to_string(),
                    s.fields.clone(),
                    None,
                    s.ident.clone(),
                    &mut impl_body,
                );

                s.attrs.retain(|a| !is_sitter_attr(a));
                s.fields.iter_mut().for_each(|f| {
                    f.attrs.retain(|a| !is_sitter_attr(a));
                });

                let struct_name = &s.ident;
                let extract_ident =
                    Ident::new(&format!("extract_{}", struct_name), Span::call_site());

                let extract_impl: Item = syn::parse_quote! {
                    impl rust_sitter::Extract for #struct_name {
                        #[allow(non_snake_case)]
                        fn extract(node: Option<rust_sitter::Node>, source: &[u8], _vec_field_name: Option<&str>) -> Self {
                            let node = node.unwrap();
                            #(#impl_body)*
                            #extract_ident(node, source)
                        }
                    }
                };

                vec![Item::Struct(s), extract_impl]
            }

            o => vec![o],
        })
        .collect();

    let tree_sitter_ident = Ident::new(&format!("tree_sitter_{}", grammar_name), Span::call_site());

    transformed.push(syn::parse_quote! {
        extern "C" {
            fn #tree_sitter_ident() -> rust_sitter::Language;
        }
    });

    transformed.push(syn::parse_quote! {
        fn language() -> rust_sitter::Language {
            unsafe { #tree_sitter_ident() }
        }
    });

    transformed.push(syn::parse_quote! {
      pub fn parse(input: &str) -> core::result::Result<#root_type, Vec<rust_sitter::errors::ParseError>> {
          let mut parser = rust_sitter::Parser::new();
          parser.set_language(language()).unwrap();
          let tree = parser.parse(input, None).unwrap();
          let root_node = tree.root_node();

          if root_node.has_error() {
              let mut errors = vec![];
              rust_sitter::errors::collect_parsing_errors(
                  &root_node,
                  &mut errors,
              );

              Err(errors)
          } else {
              use rust_sitter::Extract;
              Ok(rust_sitter::Extract::extract(Some(root_node), input.as_bytes(), None))
          }
      }
  });

    let mut filtered_attrs = input.attrs;
    filtered_attrs.retain(|a| !is_sitter_attr(a));
    ItemMod {
        attrs: filtered_attrs,
        vis: input.vis,
        mod_token: input.mod_token,
        ident: input.ident,
        content: Some((brace, transformed)),
        semi: input.semi,
    }
}
