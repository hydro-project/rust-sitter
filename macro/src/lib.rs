use quote::ToTokens;
use syn::{parse_macro_input, AttributeArgs, ItemMod};

mod expansion;
use expansion::*;

#[proc_macro_attribute]
/// Marks the top level AST node where parsing should start.
///
/// ## Example
/// ```ignore
/// #[rust_sitter::language]
/// pub struct Code {
///     ...
/// }
/// ```
pub fn language(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

#[proc_macro_attribute]
/// Marks the AST node as an extra, which can be safely skipped over while parsing.
///
/// ## Example
/// ```ignore
/// #[rust_sitter::extra]
/// pub struct Comment {
///     ...
/// }
/// ```
pub fn extra(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

#[proc_macro_attribute]
/// Defines a field which matches a specific token in the source string.
/// The token can be defined by passing one of two arguments
/// - `text`: a string literal that will be exactly matched
/// - `pattern`: a regular expression that will be matched against the source string
///
/// If the resulting token needs to be converted into a richer type at runtime,
/// such as a number, then the `transform` argument can be used to specify a function
/// that will be called with the token's text.
///
/// ## Example
/// ```ignore
/// Number(
///     #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
///     u32
/// )
/// ```
pub fn leaf(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

#[proc_macro_attribute]
/// Defines a field that does not correspond to anything in the input string,
/// such as some metadata. Takes a single, unnamed argument, which is the value
/// used to populate the field at runtime.
///
/// ## Example
/// ```ignore
/// struct MyNode {
///    ...,
///    #[rust_sitter::skip(false)]
///    node_visited: bool
/// }
/// ```
pub fn skip(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

#[proc_macro_attribute]
/// Defines a precedence level for a non-terminal that has no associativity.
///
/// This annotation takes a single, unnamed parameter, which specifies the precedence level.
/// This is used to resolve conflicts with other non-terminals, so that the one with the higher
/// precedence will bind more tightly (appear lower in the parse tree).
///
/// ## Example
/// ```ignore
/// #[rust_sitter::prec(1)]
/// PriorityExpr(Box<Expr>, Box<Expr>)
/// ```
pub fn prec(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

#[proc_macro_attribute]
/// Defines a precedence level for a non-terminal that should be left-associative.
/// For example, with subtraction we expect 1 - 2 - 3 to be parsed as (1 - 2) - 3,
/// which corresponds to a left-associativity.
///
/// This annotation takes a single, unnamed parameter, which specifies the precedence level.
/// This is used to resolve conflicts with other non-terminals, so that the one with the higher
/// precedence will bind more tightly (appear lower in the parse tree).
///
/// ## Example
/// ```ignore
/// #[rust_sitter::prec_left(1)]
/// Subtract(Box<Expr>, Box<Expr>)
/// ```
pub fn prec_left(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

#[proc_macro_attribute]
/// Defines a precedence level for a non-terminal that should be right-associative.
/// For example, with cons we could have 1 :: 2 :: 3 to be parsed as 1 :: (2 :: 3),
/// which corresponds to a right-associativity.
///
/// This annotation takes a single, unnamed parameter, which specifies the precedence level.
/// This is used to resolve conflicts with other non-terminals, so that the one with the higher
/// precedence will bind more tightly (appear lower in the parse tree).
///
/// ## Example
/// ```ignore
/// #[rust_sitter::prec_right(1)]
/// Cons(Box<Expr>, Box<Expr>)
/// ```
pub fn prec_right(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

#[proc_macro_attribute]
/// On `Vec<_>` typed fields, specifies a non-terminal that should be parsed in between the elements.
/// The [`rust_sitter::repeat`] annotation must be used on the field as well.
///
/// This annotation takes a single, unnamed argument, which specifies a field type to parse. This can
/// either be a reference to another type, or can be defined as a `leaf` field. Generally, the argument
/// is parsed using the same rules as an unnamed field of an enum variant.
///
/// ## Example
/// ```ignore
/// #[rust_sitter::delimited(
///     #[rust_sitter::leaf(text = ",")]
///     ()
/// )]
/// numbers: Vec<Number>
/// ```
pub fn delimited(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

#[proc_macro_attribute]
/// On `Vec<_>` typed fields, specifies additional config for how the repeated elements should
/// be parsed. In particular, this annotation takes the following named arguments:
/// - `non_empty` - if this argument is `true`, then there must be at least one element parsed
///
/// ## Example
/// ```ignore
/// #[rust_sitter::repeat(non_empty = true)]
/// numbers: Vec<Number>
/// ```
pub fn repeat(
    _attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    item
}

/// Mark a module to be analyzed for a Rust Sitter grammar. Takes a single, unnamed argument, which
/// specifies the name of the grammar. This name must be unique across all Rust Sitter grammars within
/// a compilation unit.
#[proc_macro_attribute]
pub fn grammar(
    attr: proc_macro::TokenStream,
    input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let attrs: AttributeArgs = parse_macro_input!(attr);
    let module: ItemMod = parse_macro_input!(input);
    let expanded: ItemMod = expand_grammar(syn::parse_quote! {
        #[rust_sitter::grammar[#(#attrs),*]]
        #module
    });
    proc_macro::TokenStream::from(expanded.to_token_stream())
}

#[cfg(test)]
mod tests {
    use std::fs::File;
    use std::io::{Read, Write};
    use std::process::Command;

    use quote::ToTokens;
    use syn::parse_quote;
    use tempfile::tempdir;

    use super::expand_grammar;

    fn rustfmt_code(code: &str) -> String {
        let dir = tempdir().unwrap();
        let file_path = dir.path().join("temp.rs");
        let mut file = File::create(file_path.clone()).unwrap();

        writeln!(file, "{code}").unwrap();
        drop(file);

        Command::new("rustfmt")
            .arg(file_path.to_str().unwrap())
            .spawn()
            .unwrap()
            .wait()
            .unwrap();

        let mut file = File::open(file_path).unwrap();
        let mut data = String::new();
        file.read_to_string(&mut data).unwrap();
        drop(file);
        dir.close().unwrap();
        data
    }

    #[test]
    fn enum_transformed_fields() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    #[rust_sitter::language]
                    pub enum Expression {
                        Number(
                            #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse::<i32>().unwrap())]
                            i32
                        ),
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn enum_recursive() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    #[rust_sitter::language]
                    pub enum Expression {
                        Number(
                            #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
                            i32
                        ),
                        Neg(
                            #[rust_sitter::leaf(text = "-")]
                            (),
                            Box<Expression>
                        ),
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn enum_prec_left() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    #[rust_sitter::language]
                    pub enum Expression {
                        Number(
                            #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
                            i32
                        ),
                        #[rust_sitter::prec_left(1)]
                        Sub(
                            Box<Expression>,
                            #[rust_sitter::leaf(text = "-")]
                            (),
                            Box<Expression>
                        ),
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn struct_extra() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    #[rust_sitter::language]
                    pub enum Expression {
                        Number(
                            #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())] i32,
                        ),
                    }

                    #[rust_sitter::extra]
                    struct Whitespace {
                        #[rust_sitter::leaf(pattern = r"\s")]
                        _whitespace: (),
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn grammar_unboxed_field() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    #[rust_sitter::language]
                    pub struct Language {
                        e: Expression,
                    }

                    pub enum Expression {
                        Number(
                            #[rust_sitter::leaf(pattern = r"\d+", transform = |v: &str| v.parse::<i32>().unwrap())]
                            i32
                        ),
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn struct_repeat() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    #[rust_sitter::language]
                    pub struct NumberList {
                        numbers: Vec<Number>,
                    }

                    pub struct Number {
                        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
                        v: i32
                    }

                    #[rust_sitter::extra]
                    struct Whitespace {
                        #[rust_sitter::leaf(pattern = r"\s")]
                        _whitespace: (),
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn struct_optional() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    #[rust_sitter::language]
                    pub struct Language {
                        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
                        v: Option<i32>,
                        t: Option<Number>,
                    }

                    pub struct Number {
                        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
                        v: i32
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn enum_with_unamed_vector() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    pub struct Number {
                            #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
                            value: u32
                    }

                    #[rust_sitter::language]
                    pub enum Expr {
                        Numbers(
                            #[rust_sitter::repeat(non_empty = true)]
                            Vec<Number>
                        )
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn enum_with_named_field() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    #[rust_sitter::language]
                    pub enum Expr {
                        Number(
                                #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
                                u32
                        ),
                        Neg {
                            #[rust_sitter::leaf(text = "!")]
                            _bang: (),
                            value: Box<Expr>,
                        }
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }

    #[test]
    fn spanned_in_vec() {
        insta::assert_display_snapshot!(rustfmt_code(
            &expand_grammar(parse_quote! {
                #[rust_sitter::grammar("test")]
                mod grammar {
                    use rust_sitter::Spanned;

                    #[rust_sitter::language]
                    pub struct NumberList {
                        numbers: Vec<Spanned<Number>>,
                    }

                    pub struct Number {
                        #[rust_sitter::leaf(pattern = r"\d+", transform = |v| v.parse().unwrap())]
                        v: i32
                    }

                    #[rust_sitter::extra]
                    struct Whitespace {
                        #[rust_sitter::leaf(pattern = r"\s")]
                        _whitespace: (),
                    }
                }
            })
            .to_token_stream()
            .to_string()
        ));
    }
}
