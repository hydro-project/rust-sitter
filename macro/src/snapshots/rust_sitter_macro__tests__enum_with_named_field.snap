---
source: macro/src/lib.rs
expression: "rustfmt_code(&expand_grammar(parse_quote! {\n                            #[rust_sitter :: grammar(\"test\")] mod grammar\n                            {\n                                #[rust_sitter :: language] pub enum Expr\n                                {\n                                    Number(#[rust_sitter ::\n                                    leaf(pattern = r\"\\d+\", transform = | v |\n                                    v.parse().unwrap())] u32), Neg\n                                    {\n                                        #[rust_sitter :: leaf(text = \"!\")] _bang : (), value : Box <\n                                        Expr >,\n                                    }\n                                }\n                            }\n                        }).to_token_stream().to_string())"
---
mod grammar {
    pub enum Expr {
        Number(u32),
        Neg { _bang: (), value: Box<Expr> },
    }
    impl rust_sitter::Extract for Expr {
        #[allow(non_snake_case)]
        fn extract(
            node: Option<rust_sitter::Node>,
            source: &[u8],
            _vec_field_name: Option<&str>,
        ) -> Self {
            let node = node.unwrap();
            #[allow(non_snake_case)]
            #[allow(clippy::unused_unit)]
            fn extract_Expr_Number_0(node: Option<rust_sitter::Node>, source: &[u8]) -> u32 {
                fn make_transform() -> impl Fn(&str) -> u32 {
                    |v| v.parse().unwrap()
                }
                make_transform()(node.and_then(|n| n.utf8_text(source).ok()).unwrap())
            }
            #[allow(non_snake_case)]
            fn extract_Expr_Number(node: rust_sitter::Node, source: &[u8]) -> Expr {
                Expr::Number(extract_Expr_Number_0(node.child_by_field_name("0"), source))
            }
            #[allow(non_snake_case)]
            #[allow(clippy::unused_unit)]
            fn extract_Expr_Neg__bang(node: Option<rust_sitter::Node>, source: &[u8]) -> () {
                rust_sitter::Extract::extract(node, source, Some("Expr_Neg__bang_vec_element"))
            }
            #[allow(non_snake_case)]
            #[allow(clippy::unused_unit)]
            fn extract_Expr_Neg_value(node: Option<rust_sitter::Node>, source: &[u8]) -> Box<Expr> {
                rust_sitter::Extract::extract(node, source, Some("Expr_Neg_value_vec_element"))
            }
            #[allow(non_snake_case)]
            fn extract_Expr_Neg(node: rust_sitter::Node, source: &[u8]) -> Expr {
                Expr::Neg {
                    _bang: extract_Expr_Neg__bang(node.child_by_field_name("_bang"), source),
                    value: extract_Expr_Neg_value(node.child_by_field_name("value"), source),
                }
            }
            match node.child(0).unwrap().kind() {
                "Expr_Number" => extract_Expr_Number(node.child(0).unwrap(), source),
                "Expr_Neg" => extract_Expr_Neg(node.child(0).unwrap(), source),
                _ => panic!(),
            }
        }
    }
    extern "C" {
        fn tree_sitter_test() -> rust_sitter::Language;
    }
    fn language() -> rust_sitter::Language {
        unsafe { tree_sitter_test() }
    }
    pub fn parse(input: &str) -> core::result::Result<Expr, Vec<rust_sitter::errors::ParseError>> {
        let mut parser = rust_sitter::Parser::new();
        parser.set_language(language()).unwrap();
        let tree = parser.parse(input, None).unwrap();
        let root_node = tree.root_node();
        if root_node.has_error() {
            let mut errors = vec![];
            rust_sitter::errors::collect_parsing_errors(&root_node, &mut errors);
            Err(errors)
        } else {
            use rust_sitter::Extract;
            Ok(rust_sitter::Extract::extract(
                Some(root_node),
                input.as_bytes(),
                None,
            ))
        }
    }
}

